File: python_refactor/hetu/optim/optimizer.py
import hetu
from hetu import Tensor
from typing import Optional, Iterable

class Optimizer(object):

    def __init__(self, params: Optional[Iterable[Tensor]], defaults: dict):
        self.params = params
        if self.params is not None:
            self.params = list(self.params)
            assert len(self.params) > 0, "No variables are provided"
            for p in self.params:
                print(type(p), p.requires_grad)
                assert p.requires_grad, f"Parameter {p} is not requires_grad"
        
        self.defaults = defaults
    
    def _assert_dar_mode(self, fn):
        if self.params is not None:
            raise ValueError(
                f"Function '{fn}' should only be used in define-and-run mode")
    
    def _assert_dbr_mode(self, fn):
        if self.params is None:
            raise ValueError(
                f"Function '{fn}' should only be used in define-by-run mode")
    
    def zero_grad(self):
        self._assert_dbr_mode('zero_grad')
        for p in self.params:
            p.zero_grad()
    
    def step(self):
        self._assert_dbr_mode('step')
        update_ops = []
        for p in self.params:
            if p.grad is None:
                continue
            update_op = self.apply_dense(p, p.grad)
            update_ops.append(update_op)
        hetu.group(update_ops).get_or_compute()
    
    def minimize(self, loss, var_list=None, name=None, grad_loss=None):
        # self._assert_dar_mode('minimize')
        update_ops = []
        grads_and_vars = self.compute_gradients(
            loss, 
            var_list=var_list, 
            grad_loss=grad_loss)
        for g, v in grads_and_vars:
            if g is None:
                continue
            update_op = self.apply_dense(v, g)
            update_ops.append(update_op)
        return hetu.group(update_ops, name=name)
    
    def compute_gradients(self, loss, var_list=None, grad_loss=None):
        if var_list is None:
            var_list = hetu.requires_grad_variables(loss)
        else:
            for v in var_list:
                assert v.requires_grad, f"Variable {v} is not requires_grad"
        # for v in var_list:
        #     print(v, " ", v.requires_grad)
        grad_list = hetu.gradients(loss, var_list, grad_loss)
        assert len(grad_list) == len(var_list), \
            f"Only {len(grad_list)} gradients are returned for " + \
            f"{len(var_list)} variables."
        return list(zip(grad_list, var_list))
    
    def apply_dense(self, param, grad):
        raise NotImplementedError

File: hetu/autograd/optim/Optimizer.cc
#include "hetu/autograd/optim/Optimizer.h"
#include "hetu/autograd/autograd.h"
#include "hetu/autograd/topo.h"
#include "hetu/autograd/ops/Optimizer.h"
#include "hetu/autograd/ops/Group.h"

namespace hetu {
namespace autograd {

Tensor Optimizer::Minimize(const Tensor& loss, const TensorList& var_list,
                           const Tensor& grad_loss, const OpName& name) {
  GradAndVarList grads_and_vars = ComputeGradients(loss, var_list, grad_loss);
  GradAndVarList filtered_grads_and_vars;
  filtered_grads_and_vars.reserve(grads_and_vars.size());
  std::copy_if(grads_and_vars.begin(), grads_and_vars.end(),
               std::back_inserter(filtered_grads_and_vars),
               [](const GradAndVar& n) { return n.first.is_defined(); });
  return ApplyGradients(filtered_grads_and_vars, name);
}

Tensor Optimizer::ApplyGradients(const GradAndVarList& grads_and_vars,
                                 const OpName& name) {
  TensorList update_edges;
  update_edges.reserve(grads_and_vars.size());
  std::transform(
    grads_and_vars.begin(), grads_and_vars.end(),
    std::back_inserter(update_edges),
    [&](const GradAndVar& grad_and_var) { return ApplyDense(grad_and_var); });
  return GroupOp(update_edges, OpMeta().set_name(name))->out_dep_linker();
}

void Optimizer::ZeroGrad() {
  for (auto& var : _vars)
    var->ZeroGrad();
}

void Optimizer::Step() {
  TensorList update_edges;
  update_edges.reserve(_vars.size());
  for (size_t i = 0; i < _vars.size(); i++) {
    auto& var = _vars[i];
    auto& grad = var->Gradient();
    if (!grad.is_defined())
      continue;
    auto update_edge = ApplyDense({grad, var});
    update_edges.push_back(update_edge);
  }
  auto update_op = GroupOp(update_edges)->out_dep_linker();
  update_op->GetOrCompute();
}

Tensor Optimizer::MakeStates(const Tensor& variable, const OpName& state_name) {
  auto var = reinterpret_cast<const VariableOp&>(variable->producer());
  return VariableOp(variable->shape(), ZerosInitializer(), variable->dtype(),
                    false,
                    OpMeta()
                      .set_device_group(var->device_group())
                      .set_name(var->name() + "_" + state_name))
    ->output(0);
}

GradAndVarList Optimizer::ComputeGradients(const Tensor& loss,
                                           const TensorList& var_list,
                                           const Tensor& grad_loss) {
  TensorList vars = var_list;
  if (vars.empty()) {
    auto topo_order = TopoSort(loss);
    vars.reserve(topo_order.size());
    for (auto& op : topo_order)
      if (is_trainable_op(op))
        vars.push_back(op->output(0));
  } else {
    for (const auto& var : vars)
      HT_ASSERT(is_trainable_op(var->producer()))
        << "Operator \"" << var->producer()->name() << "\" is not trainable";
  }
  TensorList grads = Gradients(loss, vars, grad_loss);
  HT_ASSERT_EQ(grads.size(), vars.size())
    << "Only " << grads.size() << " gradients are returned for " << vars.size()
    << " variables.";
  GradAndVarList grads_and_vars;
  grads_and_vars.reserve(grads.size());
  for (size_t i = 0; i < grads.size(); i++)
    grads_and_vars.emplace_back(grads[i], vars[i]);
  return grads_and_vars;
}

Tensor SGDOptimizer::ApplyDense(const GradAndVar& grad_and_var) {
  Tensor grad = grad_and_var.first;
  Tensor var = grad_and_var.second;
  auto update_op_meta = OpMeta()
                          .set_device_group(var->producer()->device_group())
                          .set_name("Update_" + var->name());
  if (momentum() == 0) {
    return SGDUpdateOp(var, grad, learning_rate(), update_op_meta)
      ->out_dep_linker();
  } else {
    return MomemtumUpdateOp(var, grad, MakeStates(var, "velocity"),
                            learning_rate(), momentum(), nesterov(),
                            update_op_meta)
      ->out_dep_linker();
  }
}

} // namespace autograd
} // namespace hetu
